### PR TITLE
Optimize approaches page

### PR DESCRIPTION
**Pull request by ErikSchierboom**
- Optimize approach
- Allow combining non-activerecord users
- Optimize approaches page

### CODE REVIEW

File: app/assemblers/combine_authors_and_contributors.rb [hunk 1]
```diff
@@ -4,8 +4,16 @@ class CombineAuthorsAndContributors
...
-    users = authors.random.limit(limit).to_a
-    users += contributors.random.limit(limit - users.count).to_a
+    users = randomize(authors, limit)
+    users += randomize(contributors, limit - users.count)
     users
   end
+
+  private
+  def randomize(users, limit)
+    return users.random.limit(limit).to_a if users.respond_to?(:random)
```
Comment (ErikSchierboom) [inline app/assemblers/combine_authors_and_contributors.rb:14]: This allows randomizing arrays, not only active record collections.

File: app/serializers/serialize_approaches.rb [hunk 2]
```diff
@@ -3,20 +3,62 @@ class SerializeApproaches
...
+    users.select { |user| approach_contributor_ids[approach.id]&.include?(user.id) }
+  end
+
+  memoize
+  def approach_author_ids
+    Exercise::Approach::Authorship.
+      where(approach: approaches_with_track).
+      pluck(:exercise_approach_id, :user_id).
+      group_by(&:first).
+      transform_values { |values| values.map(&:second) }
```
Comment (ErikSchierboom) [inline app/serializers/serialize_approaches.rb:30]: This build a hash from approach id to author (user) ids

File: app/assemblers/combine_authors_and_contributors.rb (hunk 1 above)
Comment (iHiD) [reply app/assemblers/combine_authors_and_contributors.rb:14]: Are both approaches in this document method used?

Comment (iHiD) [reply app/assemblers/combine_authors_and_contributors.rb:14]: (It's a nice pivot!)

**Review by iHiD — APPROVED** [review]
To my fluffy tired brain this looks good!

File: app/assemblers/combine_authors_and_contributors.rb (hunk 1 above)
Comment (ErikSchierboom) [reply app/assemblers/combine_authors_and_contributors.rb:14]: The other approach is used in `approaches#show`.

**Pull request merged**